dotnet 8 / crossplatform branch

o7Moon opened the pull request:
this has a few final bugfixes for issues that only existed on the net8 branch. the only remaining one that im aware of is the rare ogg decoding crash and the fix there will likely just be suppressing the error since it really isn't a fatal one and i think it's an issue in nvorbis rather than something lro does. 
but i think this branch is ready to merge with main and layers.

changes:
- .net framework 4.5 -> .net 8
- opentk 2.0.0 -> 4.8.2
- system.drawing -> skiasharp (due to ms removing the former as a cross platform option in newer dotnet)
- nvorbis 0.8.5 -> 0.9.1
- removed most dependency on windows forms. there is a windows build target which uses a little bit of winforms, only for things we dont have cross platform support for. this windows target will probably stay because of the way microsoft insists on handling graphical apps (have to build for a "WinExe" target on windows or else the app will create a command prompt output window that you cant get rid of without closing the app, so this separate target is specifically for that)
- updated the openal dll to 64 bit

OddMorning commented:
Sadly I can't build this on Win10 :(

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at OpenTK.Windowing.GraphicsLibraryFramework.GLFWNative.glfwCreateWindow(Int32, Int32, Byte*, OpenTK.Windowing.GraphicsLibraryFramework.Monitor*, OpenTK.Windowing.GraphicsLibraryFramework.Window*)
--------------------------------
   at OpenTK.Windowing.GraphicsLibraryFramework.GLFW.CreateWindow(Int32, Int32, System.String, OpenTK.Windowing.GraphicsLibraryFramework.Monitor*, OpenTK.Windowing.GraphicsLibraryFramework.Window*)
   at OpenTK.Windowing.Desktop.NativeWindow..ctor(OpenTK.Windowing.Desktop.NativeWindowSettings)
   at OpenTK.Windowing.Desktop.GameWindow..ctor(OpenTK.Windowing.Desktop.GameWindowSettings, OpenTK.Windowing.Desktop.NativeWindowSettings)
   at linerider.MainWindow..ctor()
   at linerider.Program.Run(System.String[])
   at linerider.EntryPoint.Main(System.String[])
```

o7Moon commented:
oh incredibly weird, ive been building on win10 a couple times. i'll go try a clean clone of the branch and see if i get the same as you

o7Moon commented:
i cloned the Experimental-Net-8 branch from my fork and built it with vs2022 on win10 and was able to run it. the only thing i can think of is, make sure its targeting 64 bit? that what it defaulted to for me but maybe not in your case.

OddMorning commented:
I asked my friend to build this branch and it works on his PC. It seems there's something wrong with my visual studio (or my PC) :\

OddMorning commented:
It seems newer OpenTK simply doesn't work in VMware (that I use for coding) or HyperV because of missing video card. Even Mesa3D didn't help (they promised to run OpenGL apps without GPU), or probably I just installed it some wrong way.
Anyway, it seems I'm the only one who tried to launch LRO this way.

OddMorning commented:
Setting `APIVersion` to `new Version(3,1)` (in `MainWindow.cs`) makes it work under VMware. Well, it fills the console with "GL Error: InvalidOperation" messages but the rest of the app seems working as expected.

Version `3.2` and above throws "render buffer not supported" error.

Can you check, does `3.1` break anything?

o7Moon commented:
3.1 works, 3.2 gets the same error as you, if you replace `ContextProfile.Any` with `ContextProfile.Compatability` then 3.2 works.
this issue i noticed in the renderer rewrite, its an issue with parsing the gl version string in `SafeFrameBuffer.cs`. it looks like i fixed it in that branch:
```csharp
string[] version = GL.GetString(StringName.Version).Split('.');
int major = int.Parse(version[0]);
// cut it off after a space in case the string looks something like "4.6 mesa (...)" like it does on my system
int minor = int.Parse(version[1].Split(' ')[0]);
_version = new Version(major, minor);
```

but i think if a 3.2 compatability context works for you then we'll use that until the renderer rewrite is ready

o7Moon commented:
i haven't had 2.1 tested on mac actually, i think maybe the current renderer can work on mac under a 2.1 context. 
so:
- if 2.1 and ContextProfile.Any works for you and works on mac then we use that
- if it doesnt work on mac and 3.2 and ContextProfile.Comapatability works for you then we'll use that (Compatibility doesnt exist on mac, which is why 2.1 would be preferred)
- otherwise we might be able to use 3.1

o7Moon commented:
I'm not sure if you ever tested 3.2 compatibility but this is probably the best gl version for us to be on rn. we should be able to gradually replace the bits that use legacy gl apis until we can remove compatibility mode entirely (since mac doesn't have it, so whenever we can remove it is when we can support mac)

OddMorning commented:
Just managed to check it. Yep, 3.2 with compatability profile works great for me. 

In general, I've noticed some minor things like silly looking track loading spinner and missing window icon, the rest of the app seems to be working as expected.

Though I'm not sure about these endless error messages, what they mean, how to trace them and whether they break anything. Probably it's just because of legacy API, right?
![mstsc – 12 01 2025; 20 20 56](https://github.com/user-attachments/assets/bc873077-7862-4ece-bb45-73816ae177c8)

o7Moon commented:
it's odd that they seem to print without gl debugging enabled for you, though i don't think i ever looked at the output on windows so maybe that just happens. i enabled debugging and got it too, its the issue cura mentioned that got fixed on the layers branch, ill add the fix here really quick.

o7Moon commented:
i don't get gl errors unless i turn on debugging in the mainwindow constructor, but i also don't know if i ever ran the windows build in a cmd to look at stdout so maybe that just happens. i think it's fixed with the above commit
